### DISCUSSION TITLE
Customizing Prompt just for Login

### DISCUSSION THREAD

**Discussion by allan-silverstein** [carlmontanari/scrapli#88] [Q&A] [answered]
Hello Carl,
  As you advised, I'm continuing this discussion here...   Regarding the "Platform Regex" section in "Advance Usage" (https://carlmontanari.github.io/scrapli/user_guide/advanced_usage/#platform-regex)

I'm little confused about this section. Specifically about using "line" anchors because all regex searches are "multi-line.  I'm assuming one large "blob" of data is received on the channel.  That data one large string which will contain newlines (\n) in the data; correct?  But I don't see "new-line" patterns in comms_prompt_pattern.  Maybe I'm not understanding something with "line anchors".  I thought those were anchors so that your regex pattern had to match the "string" in its entirety (from start to end).  Where the string itself may be comprised of multiple lines...   Unless you are saying that the scrapli code only does match checks on one line at a time.  That is how I designed TCL/Expect scripts years ago.  There would be an expect loop which would match either newline or prompt.  If it matched newline, then the buffer was reset and regex try to match on the next line.  Thus, the pattern match was only ever done one one line at a time.  

Maybe you can comment to clarify and then I can try and make some changes to get the Drivenets device to work.
Thx
Al Silverstein

This pattern is contained in the comms_prompt_pattern setting or is created by joining all possible prompt patterns in the privilege levels for a "core" device type. In general, you should not change the patterns unless you have a good reason to do so!

The "base" Driver (default, but changeable) pattern is:

"^[a-z0-9.\-@()/:]{1,48}[#>$]\s*$"

NOTE all comms_prompt_pattern "should" use the start and end of line anchors as all regex searches in scrapli are multi-line (this is an important piece to making this all work!).

**Comment by allan-silverstein** [comment]
I see in the debug output it shows here that the prompt does not start on a new line... When you log on to the box live, the string DRIVENETS CLI Loading" presents visually what appears to be a spinning star. Just changing characters between /-|\.  After about 5 seconds, the prompt appears. and the string DRIVENETS CLI Loading actually disappears...  It appears but as though the prompt "AUTTX22CWP# does begin at a new line but we know what is really happening behind the curtain. 

12    | 2021-03-02 18:43:35,162 | DEBUG    | 10.173.125.180:22         | read : b'\nDRIVENETS CLI Loading...                                                                                AUTTX22CWP# '

**Reply by carlmontanari** [reply]
Ok cool that’s good to know. Will try to respond to this tomorrow am, should be easy enough to get it going. Thanks for moving this to discussions!

**Comment by allan-silverstein** [comment]
I created the connection object but I don't see conn.channel.comms_prompt_pattern to change it...  I must be missing something.  Does it only show up after an "open"?

(Pdb) dir(conn.channel)
['__class__', '__delattr__', '__dict__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__le__', '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', '__weakref__', '_base_channel_args', '_channel_lock', '_get_prompt_pattern', '_join_and_compile', '_pre_send_input', '_pre_send_inputs_interact', '_process_output', '_read_until_input', '_read_until_prompt', '_read_until_prompt_or_time', '_ssh_message_handler', '_strip_ansi', 'channel_authenticate_ssh', 'channel_authenticate_telnet', 'channel_lock', 'channel_log', 'get_prompt', 'logger', 'read', 'send_input', 'send_input_and_read', 'send_inputs_interact', 'send_return', 'transport', 'write']
(Pdb)

**Reply by carlmontanari** [reply]
Will be just on the “conn” object. May need to update docs to reflect - was some changes in the last big update.

**Comment by allan-silverstein** [comment, marked as answer]
Ok, so I got it working just by temporarily changing comms_prompt_pattern for logging in and then immediately changing it back.  This seems simple enough.  Are there gotchas that I need to be aware of or that should suffice and I should be good (at least with logging on and sending basic commands)??  I'm sure I'll have more questions as I continue to use it..

from scrapli.driver import GenericDriver
from scrapli.logging import enable_basic_logging


```
MY_DEVICE = {
    "host": "10.173.125.180",
    "auth_username": "xxxxx",
    "auth_password": "xxxxx",
    "auth_strict_key": False,
    "channel_log": True,
}

enable_basic_logging(file=True, level="debug")

def main():
    """Simple example of connecting to an IOSXEDevice with the GenericDriver"""
    # the `GenericDriver` is a good place to start if your platform is not supported by a "core"
    #  platform drivers
    conn = GenericDriver(**MY_DEVICE)
    prompt_orig = conn.comms_prompt_pattern
    conn.comms_prompt_pattern = "CLI Loading.+#\s*$"
    conn.open()
    conn.comms_prompt_pattern = prompt_orig

    print(conn.send_command("show interface | no-more").result)
    conn.close()

if __name__ == "__main__":
    main()
```

**Reply by carlmontanari** [reply]
For `GenericDriver`, no problem at all doing what you've got here! Yeah as long as the prompts dont change wildly and there is no concept of different priv levels (like privilege exec/config/etc.) then `GenericDriver` should be good to go to send commands all day!

As an FYI -- if you were doing things w /a `NetworkDriver` (maybe you will in the future for drivenets things?) there is a [`update_privilege_levels`](https://github.com/carlmontanari/scrapli/blob/master/scrapli/driver/network/base_driver.py#L216) method that will recompile a joined privilege pattern out of all privilege levels of the driver -- this way we can always find a prompt even if we end up at the wrong priv level. So you'd need to re-call this method after updating the priv level(s).

**Comment by allan-silverstein** [comment]
I was reading up on scrapli community and I think it makes sense to just create a Drivenets driver/platform.  Can you tell me which driver you recommend for Drivnets (Generic or Network)?  It is basically a disaggregated router.  It has the following user type environment:

- There is no concept of "enable mode" like traditional cisco ios
- You need to enter the command "configuration" to go into config mode
- You need to issue a commit command for the changes to take affect.

It's funny, I created a product called the Net-Sense Automater 15 years ago using TCL/Expect.  I no longer sell it but it still works fine on Linux systems and its still being used... (www.net-sense.com) It was not open source but it had a lot of the ideas that I'm reading in your documentation along with GUI.  There was a generic_send_cmds script that accepted template files for different vendors.  The end user just had to create a template file and define prompts and "error" strings and they were on their way.  It also had/has dedicated scripts for Cisco/Juniper/Vyatta.  The windows version no longer works on anything after Windows 7.  Microsoft did some application security work under the covers and broke things...
Thx

**Reply by carlmontanari** [reply]
> I was reading up on scrapli community and I think it makes sense to just create a Drivenets driver/platform. Can you tell me which driver you recommend for Drivnets (Generic or Network)? It is basically a disaggregated router. It has the following user type environment:

Almost certainly `NetworkDriver` you can provide whatever privilege levels you want -- it can just have the two "exec" and "configuration" (or whatever you want to call them), or it could have a billion, whatever ya need!

**Comment by allan-silverstein** [comment]
Is there a way to inject my own messages into channel_log.  Hopefully, as this is important for writing out messages directly to log at exact point in time...  For example, it can be done in Netmiko using the _write_session_log method.  
connObj._write_session_log("My own embedded log message")
